astropy/wcs/src/unit_list_proxy.c

```diff
@@ -186,6 +187,46 @@ PyUnitListProxy_getitem(
   return result;
 }
 
+static PyObject*
+PyUnitListProxy_richcmp(
+	PyObject *a,
+	PyObject *b,
+	int op){
+  PyUnitListProxy *lhs, *rhs;
+  assert(a != NULL && b != NULL);
+  if (!PyObject_TypeCheck(a, &PyUnitListProxyType) || 
+  	  !PyObject_TypeCheck(b, &PyUnitListProxyType)) {
+  	Py_RETURN_NOTIMPLEMENTED;
+  }
+  if (op != Py_EQ && op != Py_NE) {
+  	Py_RETURN_NOTIMPLEMENTED;
+  }
+  lhs = (PyUnitListProxy *)a;
+  rhs = (PyUnitListProxy *)b;
+  int equal = PyObject_RichCompareBool(lhs->unit_class, rhs->unit_class, Py_EQ);
+  if (equal == -1) {
+    PyErr_SetString(PyExc_TypeError,
+                    "Some Internal Errors occured during comparison ");
+    return NULL;
+  }
+  if (op == Py_NE) {
```

Comment:
You could also combine the branches using:

```c
equal = equal == 1 && !strncmp(lhs->array, rhs->array, ARRAYSIZE) && lhs->size == rhs->size;
if ((op == Py_EQ && equal == 1) || (op == Py_NE && equal == 0)) {
  Py_RETURN_TRUE;
} else {
  Py_RETURN_FALSE;
}
```

Avoids to repeat the `strncmp` and `size` checks